### PR TITLE
`fidesctl apply` "upserts" resources

### LINKED ISSUE #279: Update `apply` to use the upsert endpoint
### Discussed in https://github.com/ethyca/fides/discussions/246

<div type='discussions-op-text'>

<sup>Originally posted by **ThomasLaPiana** November 23, 2021</sup>
The `apply` logic is currently complex due to the fact that it first checks the database for the object's existence and then does an update or create accordingly. If we switch over to using _only_ upserts, this would greatly simplify the logic.

The downside here is that we currently have pretty robust logging around when something is updated vs. created and what exactly changed (i.e. the `--diff` flag). I would argue that this functionality exists inherently in version control systems, but maybe that inclusion in the `apply` logic is very helpful.

Would love to hear more thoughts on this

fyi @ethyca/fides-control </div>

### PR DESCRIPTION
**Pull request by PSalant726**
Closes #279

### Code Changes

* [x] Revert `upsert_resources` changes such that it genuinely upserts resources
* [x] Prevent `default_organization` resources from being overwritten by default taxonomy upserts
* [x] Add an `/upsert` endpoint for each `resource_type` to fidesapi
* [x] Enable `fidesctl apply` to upsert resources

### Steps to Confirm

* [ ] Run `make cli` and ensure everything starts, no errors are logged
* [ ] Run `fidesctl apply --dry demo_resources/` and ensure things look correct compared to your local database
* [ ] Run `fidesctl apply demo_resources/` and ensure your local database is modified as expected

### Pre-Merge Checklist

* [x] All CI Pipelines Succeeded
* [x] Issue Requirements are Met

### CODE REVIEW

File: fidesctl/src/fidesctl/core/apply.py [hunk 1]
```diff
@@ -109,39 +70,40 @@ def apply(
 ) -> None:
     """
     Apply the current manifest file state to the server.
-    Excludes systems and registries.
```
Comment (PSalant726) [inline fidesctl/src/fidesctl/core/apply.py:112]: This is certainly not a true statement after these changes. I don't believe this was a true statement before these changes either. If I'm incorrect about this, and there's a good reason to skip system and registry resources, please let me know.

Comment (ThomasLaPiana) [reply fidesctl/src/fidesctl/core/apply.py:112]: this is a super legacy comment, it is incorrect

**Comment by ThomasLaPiana** [comment]
did some testing, looks good to me:

Output:

```python
root@bb44cb7bba94:/fides/fidesctl# fidesctl apply demo_resources/
Loading resource manifests from: demo_resources/
Taxonomy successfully created.
----------
Processing organization resources...
CREATED 0 organization resources.   
UPDATED 1 organization resources.   
----------
Processing registry resources...    
CREATED 0 registry resources. 
UPDATED 1 registry resources. 
----------
Processing system resources...
CREATED 0 system resources.
UPDATED 2 system resources.
----------
Processing dataset resources...
CREATED 0 dataset resources.
UPDATED 1 dataset resources.
----------
Processing data_use resources...
CREATED 0 data_use resources.
UPDATED 2 data_use resources.
----------
Processing policy resources...
CREATED 0 policy resources.
UPDATED 1 policy resources.
----------
```

```python
root@bb44cb7bba94:/fides/fidesctl# fidesctl apply --dry demo_resources/
Loading resource manifests from: demo_resources/
Taxonomy successfully created.
----------
Processing registry resources...
WOULD CREATE 0 registry resources.
WOULD UPDATE 1 registry resources.
----------
Processing organization resources...
WOULD CREATE 0 organization resources.
WOULD UPDATE 1 organization resources.
----------
Processing dataset resources...
WOULD CREATE 0 dataset resources.
WOULD UPDATE 1 dataset resources.
----------
Processing system resources...
WOULD CREATE 0 system resources.
WOULD UPDATE 2 system resources.
----------
Processing policy resources...
WOULD CREATE 0 policy resources.
WOULD UPDATE 1 policy resources.
----------
Processing data_use resources...
WOULD CREATE 0 data_use resources.
WOULD UPDATE 2 data_use resources.
----------
```

**Comment by ThomasLaPiana** [comment]
@PSalant726 when I run a non dry `apply` over and over again, it keeps saying `UPDATED 1`, but it shouldn't do anything there right?

```
root@bb44cb7bba94:/fides/fidesctl# fidesctl apply --diff demo_resources/       
Loading resource manifests from: demo_resources/
Taxonomy successfully created.
----------
Processing organization resources...
CREATED 0 organization resources.
UPDATED 1 organization resources.
----------
Processing policy resources...
CREATED 0 policy resources.
UPDATED 1 policy resources.
----------
Processing registry resources...
CREATED 0 registry resources.
UPDATED 1 registry resources.
----------
Processing system resources...
CREATED 0 system resources.
UPDATED 2 system resources.
----------
Processing data_use resources...
CREATED 0 data_use resources.
UPDATED 2 data_use resources.
----------
Processing dataset resources...
CREATED 0 dataset resources.
UPDATED 1 dataset resources.
----------
root@bb44cb7bba94:/fides/fidesctl# fidesctl apply --diff --dry demo_resources/ 
Loading resource manifests from: demo_resources/
Taxonomy successfully created.      
----------
Processing organization resources...

Updated resource with fides_key: default_organization
{}
WOULD CREATE 0 organization resources.
WOULD UPDATE 1 organization resources.
----------
Processing system resources...        

Updated resource with fides_key: demo_analytics_system
{}

Updated resource with fides_key: demo_marketing_system
{}
WOULD CREATE 0 system resources.
WOULD UPDATE 2 system resources.
----------
Processing data_use resources...

Updated resource with fides_key: third_party_sharing.legal_obligation.payroll  
{}

Updated resource with fides_key: third_party_sharing.personalized_advertising.direct_marketing
{}
WOULD CREATE 0 data_use resources.
WOULD UPDATE 2 data_use resources.
----------
Processing policy resources...

Updated resource with fides_key: demo_privacy_policy
{}
WOULD CREATE 0 policy resources.
WOULD UPDATE 1 policy resources.
----------
Processing registry resources...

Updated resource with fides_key: demo_registry      
{}
WOULD CREATE 0 registry resources.
WOULD UPDATE 1 registry resources.
----------
Processing dataset resources...

Updated resource with fides_key: demo_users_dataset
{}
WOULD CREATE 0 dataset resources.
WOULD UPDATE 1 dataset resources.
----------
```

it looks like we got rid of the `unchanged` list, so now `update` is either `unchanged or updated`?

**Comment by ThomasLaPiana** [comment]
another example:

```python
root@bb44cb7bba94:/fides/fidesctl# fidesctl apply --diff .fides/
Loading resource manifests from: .fides/
Taxonomy successfully created.
----------
Processing system resources...
CREATED 1 system resources.
UPDATED 0 system resources.
----------
Processing policy resources...
CREATED 2 policy resources.
UPDATED 0 policy resources.
----------
Processing dataset resources...
CREATED 1 dataset resources.
UPDATED 0 dataset resources.
----------
root@bb44cb7bba94:/fides/fidesctl# fidesctl apply --diff .fides/
Loading resource manifests from: .fides/
Taxonomy successfully created.
----------
Processing dataset resources...
CREATED 0 dataset resources.
UPDATED 1 dataset resources.
----------
Processing system resources...
CREATED 0 system resources.
UPDATED 1 system resources.
----------
Processing policy resources...
CREATED 0 policy resources.
UPDATED 2 policy resources.
----------
```

**Comment by ThomasLaPiana** [comment]
I guess diff also only works when its `--dry`?

**Review by ThomasLaPiana — CHANGES_REQUESTED** [review]
Found a bit of ambiguity in the way that apply communicates to the user

**Comment by PSalant726** [comment]
@ThomasLaPiana 

> when I run a non dry `apply` over and over again, it keeps saying `UPDATED 1`, but it shouldn't do anything there right?

This is because we're using a "naive" or "greedy" upsert query (`ON CONFLICT DO UPDATE`). This type of query isn't smart enough to skip an `UPDATE` if the applied change would result in an otherwise identical record. IOW, if we attempt to `INSERT` a record, and that `INSERT` would result in a database constraint violation, then an `UPDATE` is _always_ performed (even if nothing about the record changes). From the perspective of the code, this is by far the simplest way to achieve an upsert action. From the perspective of the database, this is generally inefficient, and could result in some side-effects.

<details>
<summary>What kind of side-effects?</summary>

- It might fire triggers that should not be fired.
- It write-locks "innocent" (unchanging) rows, possibly incurring costs for concurrent transactions.
- It might make the row seem new, even though it's old (transaction timestamp).
- Most importantly, with [PostgreSQL's MVCC model](https://www.postgresql.org/docs/current/mvcc-intro.html) a new row version is written for every `UPDATE`, regardless of whether the row data changed. This incurs a performance penalty for the `UPSERT` itself, table bloat, index bloat, performance penalty for subsequent operations on the table, and `VACUUM` cost. These effects are minor for queries encountering few duplicates, but massive for queries encountering many duplicates.

I went ahead with this implementation because none of those drawbacks seem to be applicable to the fidesctl database. We don't use database triggers for anything, we don't have concurrency concerns, we don't care how old records are, and even when upserting the default taxonomy over an already-persisted default taxonomy, the query still runs quickly.

</details>

<details>
<summary>What could we do differently?</summary>

**IF** we want to iterate on this approach in the future, we can create a more complex (read: smarter) upsert query like:

```sql
WITH input_rows(columns...) AS (
	VALUES (value1, value2, ...)
), ins AS (
	INSERT INTO table_name (columns...) 
	SELECT * FROM input_rows
	ON CONFLICT (targets...) DO NOTHING
	RETURNING column1, column2, ...
)
SELECT 'i' AS source                           -- 'i' for 'inserted'
     , column1, column2
FROM   ins
UNION  ALL
SELECT 's' AS source                           -- 's' for 'selected'
     , c.column1, c.column2
FROM   input_rows
JOIN   table_name c USING (targets...);
```
but I wasn't going to be able to build that with sqlalchemy in any reasonable amount of time.

The result would be a more efficient upsert action that enables us to skip rows that wouldn't receive any effective update when `ON CONFLICT` is enacted, and return to logging the "unchanged" values as we were previously.

</details>

> it looks like we got rid of the `unchanged` list, so now `update` is either `unchanged or updated`?

Unfortunately no, the `unchanged` list no longer applies because we always `UPDATE` when we don't `INSERT`.

> I guess diff also only works when its `--dry`?

This was not intended, but yes. I think I see a way of enabling `--diff` without also using `--dry`. Gonna try hacking at it for a bit.

**Comment by ThomasLaPiana** [comment]
@PSalant726 it looks like you removed the logic from `sort_create_update` that actually checked for the unchanged objects. I think adding that code back in should fix this issue. Does it seem like that'll work?

**Comment by PSalant726** [comment]
> it looks like you removed the logic from `sort_create_update` that actually checked for the unchanged objects. I think adding that code back in should fix this issue. Does it seem like that'll work?

This was intentional to avoid misleading users, although I suppose it's not always clear now either. I guess it's worth deciding if we want to 1) accurately represent what `fidesctl apply` is doing under-the-hood or 2) reassure users that nothing is changing when nothing should be changing. Regardless, reverting this change would only lead to "unchanged" returning while using `--dry`, since it's not possible to detect unchanged rows during the "real" upsert (since all rows are always _technically_ modified).

Personally, I prefer the current state of the code, and I would encourage users to use the `--diff` flag to see what's changing.

**Comment by ThomasLaPiana** [comment]
@PSalant726 while it is _technically_ an update even if it doesn't change, that isn't _really_ an update...as a user I would be really confused if I did an apply, changed nothing, did another apply and was told things were updated. I would probably assume there was a bug in either fidesctl or my manifest files.

To avoid this problem, maybe we can skip the detailed logging altogether?

```python
root@bb44cb7bba94:/fides/fidesctl# fidesctl apply demo_resources/       
Loading resource manifests from: demo_resources/
Taxonomy successfully created.
----------
Processing organization resources...
Applied 1 organization resource(s).
----------
Processing policy resources...
Applied 0 policy resource(s).
----------
Processing registry resources...
Applied 3 registry resource(s).
----------
```

and as you suggested, leave the more verbose `--diff` command in there for debugging as a user

**Review by ThomasLaPiana — APPROVED** [review]
LGTM! I'd say to make sure that the help messages in the CLI are updated but otherwise looks great, trust you to update those and ship when ready

**Pull request merged**